### PR TITLE
Handle exception while distributer not availble

### LINKED ISSUE #218: Handle the crashing of dmci if any of the distibuter crashes.
Some problems in postgres  caused dmci to crash. 
Handle these kind of exceptions due to unavailability of distributers.
And reingest the files later when the  distributers become available.

### PR DESCRIPTION
**Pull request by shamlymajeed**
**Summary**: closes #218 

**Related issue**:

**Suggested reviewer(s)**:

**Reviewer checklist**:

* [ ] The headers of all files contain a reference to the repository license
* [ ] 100% test coverage of new code - meaning:
  * [ ] The overall test coverage increased or remained the same as before
  * [ ] Every function is accompanied with a test suite
  * [ ] Tests are both positive (testing that the function work as intended with valid data) and negative (testing that the function behaves as expected with invalid data, e.g., that correct exceptions are thrown)
  * [ ] Functions with optional arguments have separate tests for all options
* [ ] Examples are supported by doctests
* [ ] All tests are passing
* [ ] All names (e.g., files, classes, functions, variables) are explicit
* [ ] Documentation (as docstrings) is complete and understandable

The checklist is based on the S-ENDA conventions and definition of done (see [General Conventions](https://s-enda-documentation.readthedocs.io/en/latest/general_conventions.html)). The above points are not necessarily relevant to all contributions. In that case, please add a short explanation to help the reviewer.

### CODE REVIEW

File: dmci/api/worker.py [hunk 1]
```diff
@@ -174,39 +174,51 @@ def distribute(self):
...
-                continue
-            obj = self.CALL_MAP[dist](
-                self._dist_cmd,
-                xml_file=self._dist_xml_file,
-                metadata_UUID=self._dist_metadata_id_uuid,
-                worker=self,
-                path_to_parent_list=self._kwargs.get(
-                    "path_to_parent_list", None),
-            )
-            valid &= obj.is_valid()
```
Comment (mortenwh) [inline dmci/api/worker.py:197]: I don't understand why we have used these obscure operators that only serve to hide the intention of the code but as far as I understand `valid` is assigned the value `True`/`False` if `obj.valid()` returns `True`/`False`. In this case, line 198 (or 205 in your version) should just be `if valid:`

File: dmci/api/worker.py [hunk 2]
```diff
@@ -174,39 +174,51 @@ def distribute(self):
...
+                )
+                valid &= obj.is_valid()
+                if obj.is_valid():
+                    obj_status, obj_msg = obj.run()
+                    status &= obj_status
+                    if obj_status:
+                        called.append(dist)
+                    else:
+                        failed.append(dist)
+                        failed_msg.append(obj_msg)
```
[the remark this answers is not in the document]
Comment (mortenwh) [reply dmci/api/worker.py:212]: it seems to be line 196 that should be "tried"?

File: dmci/api/worker.py [hunk 3]
```diff
@@ -174,39 +174,51 @@ def distribute(self):
...
                 else:
-                    failed.append(dist)
-                    failed_msg.append(obj_msg)
-            else:
-                skipped.append(dist)
-
-        return status, valid, called, failed, skipped, failed_msg
+                    skipped.append(dist)
+            except Exception as e:
+                logger.critical(" %s distributer currently not available." % dist)
```
Comment (mortenwh) [inline dmci/api/worker.py:216]:
```suggestion
                logger.critical(" %s distributor currently not available." % dist)
```

**Review by mortenwh — CHANGES_REQUESTED** [review]
We need to wok on the `distribute` function. See comments. Also, the errors seem to occur in `dmci/distributors/distributor.py`. As such, the error handling should be done there.

File: dmci/api/worker.py [hunk 4]
```diff
@@ -174,39 +174,51 @@ def distribute(self):
             The names of all distributors that were skipped (invalid)
         failed_msg : list of str
             The messages returned by the failed jobs
+        retry_status : bool
```
[the remark this answers is not in the document]
Comment (shamlymajeed) [reply dmci/api/worker.py:177]: re_send_status :bool
 True Resend if any one distributor not available.

File: dmci/api/worker.py [hunk 5]
```diff
@@ -174,39 +174,51 @@ def distribute(self):
...
+             True if any one distributer fails to process or not available
+        retry : list of str
```
[the remark this answers is not in the document]
Comment (shamlymajeed) [reply dmci/api/worker.py:179]: we can choose variable name 'unavailable'

File: dmci/api/worker.py (hunk 1 above)
Comment (shamlymajeed) [reply dmci/api/worker.py:197]: in this case ,if any of the 'obj is not valid ,the 'valid, variable is set to false.This is just used for logging  purpose.

 But we need the validity of the current 'obj' to decide whether to call or not call 'obj.run()'. So the line 198 should be as it is now.

File: dmci/api/worker.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mortenwh) [reply dmci/api/worker.py:177]: My point is that the variable name should not be given by the action you want to perform but the actual problem at hand, in this case that the distributor is not available.

File: dmci/api/worker.py (hunk 1 above)
Comment (mortenwh) [reply dmci/api/worker.py:197]: aha - ok

File: dmci/distributors/solr_dist.py [hunk 6]
```diff
@@ -101,32 +101,37 @@ def _add(self):
...
+                             newdoc['metadata_identifier'])
+                status, msg = self._index_record(
+                    newdoc, add_thumbnail=False, level=1)
         else:
-            logger.debug("Parent/Level-1 - dataset - %s",
-                         newdoc['metadata_identifier'])
-            status, msg = self._index_record(
-                newdoc, add_thumbnail=False, level=1)
-
+            msg = "Failed to insert."
```
Comment (mortenwh) [inline dmci/distributors/solr_dist.py:132]: We could perhaps be more explicit:
```suggestion
            msg = "Failed to insert dataset in SolR."
```
But when does this happen?

File: tests/test_api/test_app.py [hunk 7]
```diff
@@ -192,8 +192,10 @@ def testApiApp_InsertUpdateRequests(client, monkeypatch):
         f = ["A", "B"]
         s = ["C"]
         e = ["Reason A", "Reason B"]
+
```
Comment (mortenwh) [inline tests/test_api/test_app.py:195]:
```suggestion
```

File: tests/test_api/test_app.py [hunk 8]
```diff
@@ -192,8 +192,10 @@ def testApiApp_InsertUpdateRequests(client, monkeypatch):
...
         mp.setattr("dmci.api.app.Worker.validate", lambda *a: (True, "", MOCK_XML))
-        mp.setattr("dmci.api.app.Worker.distribute", lambda *a: (False, False, [], f, s, e))
+        mp.setattr("dmci.api.app.Worker.distribute", lambda *a: (
+                   False, False, [], f, s, e))
```
Comment (mortenwh) [inline tests/test_api/test_app.py:198]:
```suggestion
        mp.setattr("dmci.api.app.Worker.distribute", lambda *a: (False, False, [], f, s, e))
```

File: tests/test_api/test_app.py [hunk 9]
```diff
@@ -252,7 +264,8 @@ def testApiApp_DeleteRequests(client, monkeypatch):
         f = ["A", "B"]
         s = ["C"]
         e = ["Reason A", "Reason B"]
-        mp.setattr("dmci.api.app.Worker.distribute", lambda *a: (False, False, [], f, s, e))
+        mp.setattr("dmci.api.app.Worker.distribute", lambda *a: (
+                   False, False, [], f, s, e))
```
Comment (mortenwh) [inline tests/test_api/test_app.py:268]:
```suggestion
        mp.setattr("dmci.api.app.Worker.distribute", lambda *a: (False, False, [], f, s, e))
```

File: tests/test_api/test_app.py [hunk 10]
```diff
@@ -265,7 +278,8 @@ def testApiApp_DeleteRequests(client, monkeypatch):
 
     # Distribute ok
     with monkeypatch.context() as mp:
-        mp.setattr("dmci.api.app.Worker.distribute", lambda *a: (True, True, [], [], [], []))
+        mp.setattr("dmci.api.app.Worker.distribute", lambda *a: (
+                   True, True, [], [], [], []))
```
Comment (mortenwh) [inline tests/test_api/test_app.py:282]:
```suggestion
        mp.setattr("dmci.api.app.Worker.distribute", lambda *a: (True, True, [], [], [], []))
```

File: tests/test_dist/test_pycsw_dist.py [hunk 11]
```diff
@@ -81,11 +85,19 @@ def testDistPyCSW_Insert(monkeypatch, mockXml, mockXslt):
             "dmci.distributors.pycsw_dist.requests.post", lambda *a, **k: mockResp
         )
         mp.setattr(PyCSWDist, "_get_transaction_status", lambda *a: False)
+
         tstPyCSW = PyCSWDist("insert", xml_file=mockXml)
         tstPyCSW._conf.mmd_xsl_path = mockXslt
         assert tstPyCSW.run() == (False, "Mock response")
 
-# END Test testDistPyCSW_Insert
+    # Insert returns False if post raise exception
```
Comment (mortenwh) [inline tests/test_dist/test_pycsw_dist.py:89]:
```suggestion
    # Insert returns False if the http post request fails
```

File: tests/test_dist/test_pycsw_dist.py [hunk 12]
```diff
@@ -121,6 +133,12 @@ def testDistPyCSW_Update(monkeypatch, mockXml, mockXslt, tmpUUID):
         tstPyCSW._conf.mmd_xsl_path = mockXslt
         assert tstPyCSW.run() == (False, "Mock response")
 
+    # Update returns False if post raise exception
```
Comment (mortenwh) [inline tests/test_dist/test_pycsw_dist.py:132]:
```suggestion
    # Update returns False if the http post request fails
```

File: tests/test_dist/test_pycsw_dist.py [hunk 13]
```diff
@@ -154,6 +172,12 @@ def testDistPyCSW_Delete(monkeypatch, mockXml, tmpUUID):
         res = PyCSWDist("delete", metadata_UUID=tmpUUID, worker=mockWorker).run()
         assert res == (False, "Mock response")
 
+    # Delete returns False if post raise exception
```
Comment (mortenwh) [inline tests/test_dist/test_pycsw_dist.py:171]:
```suggestion
    # Delete returns False if http post request fails
```

**Review by mortenwh — CHANGES_REQUESTED** [review]
Some changes needed..

**Review by mortenwh — APPROVED** [review]

**Pull request merged**